---
title: "R_Assignment"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##import fang and snp files using the right formate for different types of variables
Abby: This looks like it worked great. I would not use chromosome as a number because some values are unknown or multiple.
```{r}
library(tidyverse)
fang <- read_tsv("./fang_et_al_genotypes.txt",col_names=FALSE)
snp <- read_tsv("./snp_position.txt",col_names=TRUE,cols(SNP_ID = col_character(),
  cdv_marker_id = col_double(),
  Chromosome = col_character(),
  Position = col_number(),
  alt_pos = col_character(),
  mult_positions = col_character(),
  amplicon = col_character(),
  cdv_map_feature.name = col_character(),
  gene = col_character(),
  `candidate/random` = col_character(),
  Genaissance_daa_id = col_double(),
  Sequenom_daa_id = col_double(),
  count_amplicons = col_double(),
  count_cmf = col_double(),
  count_gene = col_double()))
```

##inspection of files
Abby: another command you could use is names(). This gives the names of the various variables in each data frame.
```{r}
file.info("fang_et_al_genotypes.txt")
file.info("snp_position.txt")
str(fang)
str(snp)
is.data.frame(fang)
is.data.frame(snp)
nrow(fang)
nrow(snp)
ncol(fang)
ncol(snp)
names(fang)
names(snp)
```
##select out useful columns
```{r}
fang_sub <- fang[ 0:2783 , 3:986]
snp_sub <- select (snp, SNP_ID, Chromosome, Position)
maize <- filter(fang_sub,`X3` %in% c("Group","ZMMIL", "ZMMMR" ,"ZMMLR"))
teosinte <- filter(fang_sub, `X3` %in% c("Group","ZMPBA","ZMPIL","ZMPJA"))
maize <- maize[ , 2:984]
teosinte <- teosinte[ ,2:984]
```
##transpose data, as data.frame instead of matrix, specify stringsasfactors=false, strings will maintain as strings
```{r}
maize_transposed <- as.data.frame(t(maize),stringsAsFactors = FALSE)
teosinte_transposed <- as.data.frame(t(teosinte),stringsAsFactors = FALSE)
names(maize_transposed)[1]<-"SNP_ID"
names(teosinte_transposed)[1]<-"SNP_ID"
```
##join the two files together based on the common columns
```{r}
maize_joined <- merge(snp_sub,maize_transposed)
teosinte_joined <- merge(snp_sub,teosinte_transposed)

```
##sort by increasing chromosome position values for maize dataset

Abby: You could combine the filtering and the table writing in one step using the pipe. As follows:
```{r}
maize_joined %>% arrange(Chromosome, Position) %>% filter(Chromosome == 1) %>% write.table("maize_chr1_for.txt", sep="\t",col.names = FALSE, row.names = FALSE, quote = FALSE)
```


```{r}
maize_increasing <- arrange(maize_joined,Position)
maize_chr1 <- filter(maize_increasing, `Chromosome`==1)
maize_chr2 <- filter(maize_increasing, `Chromosome`==2)
maize_chr3 <- filter(maize_increasing, `Chromosome`==3)
maize_chr4 <- filter(maize_increasing, `Chromosome`==4)
maize_chr5 <- filter(maize_increasing, `Chromosome`==5)
maize_chr6 <- filter(maize_increasing, `Chromosome`==6)
maize_chr7 <- filter(maize_increasing, `Chromosome`==7)
maize_chr8 <- filter(maize_increasing, `Chromosome`==8)
maize_chr9 <- filter(maize_increasing, `Chromosome`==9)
maize_chr10 <- filter(maize_increasing, `Chromosome`==10)

write.table(maize_chr1, file="./Maize/maize_chr1_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr2, file="./Maize/maize_chr2_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr3, file="./Maize/maize_chr3_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr4, file="./Maize/maize_chr4_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr5, file="./Maize/maize_chr5_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr6, file="./Maize/maize_chr6_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr7, file="./Maize/maize_chr7_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr8, file="./Maize/maize_chr8_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr9, file="./Maize/maize_chr9_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr10, file="./Maize/maize_chr10_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)

teosinte_increasing <- arrange(teosinte_joined, Position)
teosinte_chr1 <- filter(teosinte_increasing, `Chromosome`==1)
teosinte_chr2 <- filter(teosinte_increasing, `Chromosome`==2)
teosinte_chr3 <- filter(teosinte_increasing, `Chromosome`==3)
teosinte_chr4 <- filter(teosinte_increasing, `Chromosome`==4)
teosinte_chr5 <- filter(teosinte_increasing, `Chromosome`==5)
teosinte_chr6 <- filter(teosinte_increasing, `Chromosome`==6)
teosinte_chr7 <- filter(teosinte_increasing, `Chromosome`==7)
teosinte_chr8 <- filter(teosinte_increasing, `Chromosome`==8)
teosinte_chr9 <- filter(teosinte_increasing, `Chromosome`==9)
teosinte_chr10 <- filter(teosinte_increasing, `Chromosome`==10)

write.table(teosinte_chr1, file="./teosinte/teosinte_chr1_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr2, file="./teosinte/teosinte_chr2_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr3, file="./teosinte/teosinte_chr3_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr4, file="./teosinte/teosinte_chr4_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr5, file="./teosinte/teosinte_chr5_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr6, file="./teosinte/teosinte_chr6_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr7, file="./teosinte/teosinte_chr7_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr8, file="./teosinte/teosinte_chr8_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr9, file="./teosinte/teosinte_chr9_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr10, file="./teosinte/teosinte_chr10_increasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)




```
##replace missing values by -/-
Abby: This looks like it worked. I also used three lines to accomplish the same thing.
```{r}

maize_alt <- as.matrix(maize_joined)
maize_alt <- lapply(maize_joined,gsub,pattern = "?", replacement = "-", fixed = TRUE)
maize_alt <- as_tibble(maize_alt)
maize_alt$Position <- as.numeric(maize_alt$Position)

teosinte_alt <- as.matrix(teosinte_joined)
teosinte_alt <- lapply(teosinte_joined, gsub, pattern = "?", replacement = "-", fixed = TRUE)
teosinte_alt <- as_tibble(teosinte_alt)
teosinte_alt$Position<- as.numeric(teosinte_alt$Position)

```
##sort by decreasing value 
```{r}
teosinte_decreasing <- arrange(teosinte_alt,-Position)
maize_decreasing <- arrange(maize_alt, -Position)
```
##sort by choromosome value for teosinte/maize dataset
```{r}

teosinte_chr1 <- filter(teosinte_decreasing,`Chromosome`==1)
teosinte_chr2 <- filter(teosinte_decreasing,`Chromosome`==2)
teosinte_chr3 <- filter(teosinte_decreasing,`Chromosome`==3)
teosinte_chr4 <- filter(teosinte_decreasing,`Chromosome`==4)
teosinte_chr5 <- filter(teosinte_decreasing,`Chromosome`==5)
teosinte_chr6 <- filter(teosinte_decreasing,`Chromosome`==6)
teosinte_chr7 <- filter(teosinte_decreasing,`Chromosome`==7)
teosinte_chr8 <- filter(teosinte_decreasing,`Chromosome`==8)
teosinte_chr9 <- filter(teosinte_decreasing,`Chromosome`==9)
teosinte_chr10 <- filter(teosinte_decreasing,`Chromosome`==10)

maize_chr1 <- filter(maize_decreasing, `Chromosome`==1)
maize_chr2 <- filter(maize_decreasing, `Chromosome`==2)
maize_chr3 <- filter(maize_decreasing, `Chromosome`==3)
maize_chr4 <- filter(maize_decreasing, `Chromosome`==4)
maize_chr5 <- filter(maize_decreasing, `Chromosome`==5)
maize_chr6 <- filter(maize_decreasing, `Chromosome`==6)
maize_chr7 <- filter(maize_decreasing, `Chromosome`==7)
maize_chr8 <- filter(maize_decreasing, `Chromosome`==8)
maize_chr9 <- filter(maize_decreasing, `Chromosome`==9)
maize_chr10 <- filter(maize_decreasing, `Chromosome`==10)

write.table(maize_chr1, file="./Maize/maize_chr1_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr2, file="./Maize/maize_chr2_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr3, file="./Maize/maize_chr3_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr4, file="./Maize/maize_chr4_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr5, file="./Maize/maize_chr5_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr6, file="./Maize/maize_chr6_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr7, file="./Maize/maize_chr7_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr8, file="./Maize/maize_chr8_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr9, file="./Maize/maize_chr9_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(maize_chr10, file="./Maize/maize_chr10_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)

write.table(teosinte_chr1, file="./teosinte/teosinte_chr1_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr2, file="./teosinte/teosinte_chr2_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr3, file="./teosinte/teosinte_chr3_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr4, file="./teosinte/teosinte_chr4_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr5, file="./teosinte/teosinte_chr5_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr6, file="./teosinte/teosinte_chr6_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr7, file="./teosinte/teosinte_chr7_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr8, file="./teosinte/teosinte_chr8_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr9, file="./teosinte/teosinte_chr9_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)
write.table(teosinte_chr10, file="./teosinte/teosinte_chr10_decreasing.txt", quote=FALSE, sep="\t", row.names=FALSE, col.names=TRUE)


```

##snps per chromosome & distribution for each chromosome
Abby: I would suggest changing the first plot to account for the unknown and multiple chromosome data.Remember that I did not import the chromosome list as a number. Changing the bins to 12, selecting stat = "count", and using as.character(Chromosome) helps with this. In the density plot, there will be a facet for NA now.
```{r}
fang_sub1 <- fang[0:2783 , 4:986]
fang_trans1 <- as.data.frame(t(fang_sub1),stringsAsFactors=TRUE)
names(fang_trans1)[1]<-"SNP_ID"
fang_merged1 <- merge(snp_sub,fang_trans1)
ggplot(fang_merged1,aes(x=as.character(Chromosome))) + geom_histogram(bins=12, stat = "count", aes(fill = Chromosome))

ggplot(fang_merged1, aes(x=as.numeric(Position))) +
  geom_density(aes(fill = as.character(Chromosome))) + facet_wrap(~
  as.numeric(Chromosome)) + theme_bw()
```
##creating graphs based on homozygous/heterozygous(melt first)
Abby: You can go ahead and use homozygous, missing, and heterozygous in the commands to make the graph more clear.
```{r}
library(reshape2)
fang_withHeader <- read_tsv("./fang_et_al_genotypes.txt")
fang_melted <- melt(fang_withHeader,id.var ="Sample_ID",measure.vars=names(fang_withHeader[4:986]))
new <- fang_melted %>% as.matrix()
new [new %in% c("A/A","T/T","C/C","G/G")] <- "homozygous"
new [new %in% c("A/G","A/T","C/G","C/T","G/T","A/C")] <- "heterozygous"
new [new %in% c("?/?")] <- "missing"
new <- as.data.frame(new, stringsAsFactors = FALSE)
new %>% ggplot() + geom_bar(aes(x = Sample_ID, fill=value), position = "fill")
```

##byGroup
Abby: Same thing for this plot.
```{r}
fang_melted2<- melt(fang_withHeader,id.var = "Group",measure.vars = names(fang_withHeader[4:986]))
new2 <- fang_melted2 %>% as.matrix()
new2 [new2 %in% c("A/A","T/T","C/C","G/G")] <- "homozygous"
new2 [new2 %in% c("A/G","A/T","C/G","C/T","G/T","A/C")] <- "heterozygous"
new2 [new2 %in% c("?/?")] <- "missing"
new2 <- as.data.frame(new2,stringsAsFactors =FALSE)
new2 %>% ggplot() + geom_bar(aes(x = Group, fill=value), position = "fill")
```

##visualize another feature of the dataset: Visualize the distribution of the snps on each different chromosomes. 
```{r}
ggplot(data=snp) + geom_point(mapping=aes(x=Chromosome,y=Position), color="blue")
```
